Source/FieldSolver/ImplicitSolvers/SemiImplicitEM.H

```diff
@@ -0,0 +1,121 @@
+/* Copyright 2024 Justin Angus
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#ifndef SEMI_IMPLICIT_EM_H_
+#define SEMI_IMPLICIT_EM_H_
+
+#include "FieldSolver/ImplicitSolvers/WarpXSolverVec.H"
+#include "NonlinearSolvers/NonlinearSolverLibrary.H"
+
+#include <AMReX_Array.H>
+#include <AMReX_MultiFab.H>
+#include <AMReX_REAL.H>
+
+#include "ImplicitSolver.H"
+
+/** @file
+ *  Semi-implicit electromagnetic time solver class. The electric field and the
+ *  particles are implicitly coupled in this algorithm, but the magnetic field
+ *  is advanced in the standard explicit leap-frog manner (hence semi-implicit).
+ *
+ *  The time stencil is
+ *  Eg^{n+1} = Eg^n + c^2*dt*( curlBg^{n+1/2} - mu0*Jg^{n+1/2} )
+ *  Bg^{n+3/2} = Bg^{n+1/2} - dt*curlEg^{n+1}
+ *  xp^{n+1} = xp^n + dt*up^{n+1/2}/(0.5(gammap^n + gammap^{n+1}))
```

Comment:
Equivalent but slightly shorter
```suggestion
 *  xp^{n+1} = xp^n + dt*up^{n+1/2}/gammap^{n+1/2}
```